scripts-available/CDB_SyncTable.sql

```diff
@@ -0,0 +1,150 @@
+/*
+   Gets the column names of a given table.
+
+   Sample usage:
+
+     SELECT _CDB_GetColumns('public.films');
+*/
+CREATE OR REPLACE FUNCTION _CDB_GetColumns(src_table REGCLASS)
+RETURNS SETOF NAME
+AS $$
+  SELECT
+    a.attname as "colname"
+  FROM
+    pg_catalog.pg_attribute a
+  WHERE
+    a.attnum > 0
+      AND NOT a.attisdropped
+      AND a.attrelid = (
+        SELECT c.oid
+          FROM pg_catalog.pg_class c
+          LEFT JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace
+          WHERE c.oid = src_table::oid
+            AND pg_catalog.pg_table_is_visible(c.oid)
+      );
+$$ LANGUAGE sql STABLE PARALLEL UNSAFE;
+
+
+/*
+    Given an array of quoted column names, it generates an UPDATE SET
+    clause with the following form:
+
+        the_geom = changed.the_geom,
+        id = changed.id,
+        elevation = changed.elevation
+
+    Example of usage:
+
+       SELECT __CDB_GetUpdateSetClause('{the_geom, id, elevation}', 'changed');
+*/
+CREATE OR REPLACE FUNCTION __CDB_GetUpdateSetClause(colnames TEXT[], update_source TEXT)
+RETURNS TEXT
+AS $$
+DECLARE
+  set_clause_list TEXT[];
+  col TEXT;
+BEGIN
+  FOREACH col IN ARRAY colnames
+  LOOP
+    set_clause_list := array_append(set_clause_list, format('%1$s = %2$s.%1$s', col, update_source));
+  END lOOP;
+  RETURN array_to_string(set_clause_list, ', ');
+END;
+$$ LANGUAGE plpgsql IMMUTABLE PARALLEL SAFE;
+
+/*
+   A Table Syncer
+
+   Assumptions:
+     - Both tables contain a consistent cartodb_id column
+     - Destination table has all columns of the source
+
+   Sample usage:
+
+     SELECT CDB_SyncTable('radar_stations', 'public', 'syncdest');
+
+*/
+CREATE OR REPLACE FUNCTION CDB_SyncTable(src_table REGCLASS, dst_schema REGNAMESPACE, dst_table NAME)
+RETURNS void
+AS $$
+DECLARE
+  fq_dest_table TEXT;
+
+  colnames TEXT[];
+  quoted_colnames TEXT;
+
+  src_hash_table_name NAME;
+  dst_hash_table_name NAME;
+
+  update_set_clause TEXT;
+
+  num_rows BIGINT;
+  err_context text;
+BEGIN
+  -- If the destination table does not exist, just copy the source table
+  fq_dest_table := format('%I.%I', dst_schema, dst_table);
+  EXECUTE format('CREATE TABLE IF NOT EXISTS %s as TABLE %I', fq_dest_table, src_table);
+  GET DIAGNOSTICS num_rows = ROW_COUNT;
+  IF num_rows > 0 THEN
+    RAISE NOTICE 'INSERTED % row(s)', num_rows;
+    RETURN;
+  END IF;
+
+  -- Get the list of columns from the source table, excluding cartodb_id
+  SELECT ARRAY(SELECT quote_ident(c) FROM _CDB_GetColumns(src_table) as c WHERE c <> 'cartodb_id') INTO colnames;
+  quoted_colnames := array_to_string(colnames, ',');
+
+  src_hash_table_name := format('src_sync_%s', txid_current());
+  dst_hash_table_name := format('dst_sync_%s', txid_current());
+
+  EXECUTE format('CREATE TEMP TABLE %I(cartodb_id BIGINT, hash TEXT) ON COMMIT DROP', src_hash_table_name);
```

Comment:
It could be very messy, but have you considered persisting the source hash table  (to be used as the next destination hash table)?

Reply:
Oh, I realize it's a very bad idea, regarding concurrent changes etc.

Reply:
Really not worth it, as per the few "load tests" I performed in my machine. The whole process can take less than 30s for a 1.5M rows table with changes in 10% of the rows, so... Also, the whole process is transactional because of the way PG functions work:

Up to PG10:

https://www.postgresql.org/docs/10/plpgsql-structure.html

> Functions and trigger procedures are always executed within a transaction established by an outer query — they cannot start or commit that transaction, since there would be no context for them to execute in

In PG11:

https://www.postgresql.org/about/news/1894/

> PostgreSQL 11 adds SQL procedures that can perform full transaction management within the body of a function, enabling developers to create more advanced server-side applications, such as ones involving incremental bulk data loading.

https://www.postgresql.org/docs/11/plpgsql-transactions.html